Handle alternate NT path forms

klightspeed opened the pull request:
## Summary

* OS: Windows
* Bug fix: yes
* Type: core
* Fixes: #2494

## Description

`convert_dos_path` does not currently handle UNC paths or paths of the form `\??\X:` (as returned by Wine). This results in the `psutil/tests/test_process.py::TestProcess::test_exe` test failing in the following cases:
* Python is run from a network share; or
* The tests are performed under Wine on a drive other than the one Python is installed on.

Add handling for the following NT path forms:
* `\\server\share`
* `\Device\Mup\server\share` -> `\\server\share`
* `\??\UNC\server\share` -> `\\server\share`
* `\??\X:` -> `X:`

klightspeed committed:
Handle alternate NT path forms

`convert_dos_path` does not currently handle UNC paths or paths of the form `\??\X:` (as returned by Wine).
This results in the `psutil/tests/test_process.py::TestProcess::test_exe` test failing in the following cases:
* Python is run from a network share; or
* The tests are performed under Wine on a drive other than the one Python is installed on.

Add handling for the following NT path forms:
* `\\server\share`
* `\Device\Mup\server\share` -> `\\server\share`
* `\??\UNC\server\share` -> `\\server\share`
* `\??\X:` -> `X:`

Signed-off-by: Ben Peddell <klightspeed@killerwolves.net>

giampaolo commented on psutil/_pswindows.py:
```diff
@@ -184,12 +184,21 @@ class IOPriority(enum.IntEnum):
     into:
         "C:\Windows\systemew\file.txt".
     """
+    if s[:2] == '\\\\':
```
```diff
-   if s[:2] == '\\\\':
+   if s.startswith('\\\\'):
```

giampaolo reviewed (changes requested):
Please write some tests for this. You can put in test_windows.py, `TestSystemAPIs` class.

klightspeed commented:
> Please write some tests for this. You can put in test_windows.py, `TestSystemAPIs` class.

I think any such tests would have to use [`RtlDosPathNameToNtPathName_U_WithStatus`](https://learn.microsoft.com/en-us/windows/win32/devnotes/rtldospathnametontpathname_u_withstatus) to convert various DOS paths to NT paths (and perhaps use `NtOpenSymbolicLinkObject` and `NtQuerySymbolicLinkObject` in additional tests to convert that path to one that is currently resolved by `convert_dos_path` for drives), then test that `convert_dos_path` is able to round-trip the conversion - refer [Project Zero: The Definitive Guide on Win32 to NT Path Conversion](https://googleprojectzero.blogspot.com/2016/02/the-definitive-guide-on-win32-to-nt.html) for the various DOS to NT path conversions.

Then perhaps tests would need to be written to test that `RtlDosPathNameToNtPathName_U_WithStatus` returns the expected outputs, and perhaps that `RtlDosPathNameToNtPathName_U_WithStatus` -> `NtQuerySymbolicLinkObject` returns the same path as `psutil_proc_exe` (before being converted to a DOS path by `convert_dos_path`)

And perhaps have the `psutil.proc_exe` tests use `os.path.realpath` to fix the errors when running python from e.g. a `subst`d drive or a mapped network drive.

klightspeed committed:
Add tests for _pswindows.convert_dos_path

* Test that the following return r"C:\Windows\Temp" (where
  r"\Device\HarddiskVolume1" is mounted on "C:"):
  - convert_dos_path(r"\??\C:\Windows\Temp")
  - convert_dos_path(r"\Device\HarddiskVolume1\Windows\Temp")
* Test that the following return r"\\localhost\C$\Windows\Temp":
  - convert_dos_path(r"\\localhost\C$\Windows\Temp")
  - convert_dos_path(r"\??\UNC\localhost\C$\Windows\Temp")
  - convert_dos_path(r"\Device\Mup\localhost\C$\Windows\Temp")

Signed-off-by: Ben Peddell <klightspeed@killerwolves.net>

klightspeed commented:
Drives:
```
(psutil) D:\venv\psutil\psutil>subst X: D:\venv

(psutil) D:\venv\psutil\psutil>net use Y: \\localhost\D$
The command completed successfully.


(psutil) D:\venv\psutil\psutil>python
Python 3.13.1 (tags/v3.13.1:0671451, Dec  3 2024, 19:06:28) [MSC v.1942 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import psutil
>>> import os
>>> [(x.mountpoint, os.path.realpath(x.mountpoint), x.opts) for x in psutil.disk_partitions(all=True)]
[('C:\\', 'C:\\', 'rw,fixed'), ('D:\\', 'D:\\', 'rw,fixed'), ('G:\\', 'G:\\', 'rw,fixed'), ('X:\\', 'D:\\venv', 'rw,fixed'), ('Y:\\', '\\\\localhost\\D$', 'rw,remote')]
```
Test results for these tests on my laptop:
```
(psutil) D:\venv\psutil\psutil>python -m pytest -v "psutil\tests\test_windows.py::TestSystemAPIs::test_convert_dos_path_drive" "psutil\tests\test_windows.py::TestSystemAPIs::test_convert_dos_path_unc"
======================================== test session starts ========================================
collected 2 items

psutil/tests/test_windows.py::TestSystemAPIs::test_convert_dos_path_drive PASSED
psutil/tests/test_windows.py::TestSystemAPIs::test_convert_dos_path_unc PASSED

========================================= 2 passed in 0.44s =========================================
```
External tests using python interactive:
```
>>> import psutil
>>> import os
>>> partitions = [
...     x.mountpoint
...     for x in psutil.disk_partitions(all=True)
...     if 'fixed' in x.opts or 'remote' in x.opts
... ]
>>> cext = psutil._pswindows.cext
>>> for winpath in partitions:
...     print(f'Path: {winpath!r}')
...     realpath = os.path.realpath(winpath)
...     print(f'os.path.realpath({winpath!r}) -> {realpath!r}')
...     ntpath1 = '\\??\\' + winpath
...     ntpath2 = cext.RtlDosPathNameToNtPathName(winpath)
...     print(f'cext.RtlDosPathNameToNtPathName({winpath!r}) -> {ntpath2!r}')
...     assert ntpath2 == ntpath1
...     cntpath1 = psutil._pswindows.convert_dos_path(ntpath1)
...     print(f'psutil._pswindows.convert_dos_path({ntpath1!r}) -> {cntpath1!r}')
...     assert cntpath1 == winpath
...     ntpath3 = cext.GetFinalPathName(winpath, to_nt=True)
...     print(f'cext.GetFinalPathName({winpath!r}, to_nt=True) -> {ntpath3!r}')
...
...     if realpath.startswith('\\\\'):
...         ntpath1a = '\\??\\UNC\\' + realpath[2:]
...         ntpath2a = cext.RtlDosPathNameToNtPathName(realpath)
...         print(f'cext.RtlDosPathNameToNtPathName({realpath!r}) -> {ntpath2a!r}')
...         assert ntpath2a == ntpath1a
...         cntpath1a = psutil._pswindows.convert_dos_path(ntpath1a)
...         print(f'psutil._pswindows.convert_dos_path({ntpath1a!r}) -> {cntpath1a!r}')
...         assert cntpath1a == realpath
...         ntpath4a = cext.GetFinalPathName(ntpath1a, from_nt=True, to_nt=True)
...         print(f'cext.GetFinalPathName({ntpath1a!r}, from_nt=True, to_nt=True) -> {ntpath4a!r}')
...         assert ntpath4a == ntpath3
...
...     ntpath4 = cext.GetFinalPathName(ntpath1, from_nt=True, to_nt=True)
...     print(f'cext.GetFinalPathName({ntpath1!r}, from_nt=True, to_nt=True) -> {ntpath4!r}')
...     assert ntpath4 == ntpath3
...     assert psutil._pswindows.convert_dos_path(ntpath3) == realpath
...     winpath2 = cext.GetFinalPathName(ntpath1, from_nt=True)
...     print(f'cext.GetFinalPathName({ntpath1!r}, from_nt=True) -> {winpath2!r}')
...
...     if realpath.startswith('\\\\'):
...         assert winpath2 == '\\\\?\\UNC\\' + realpath[2:]
...     else:
...         assert winpath2 == '\\\\?\\' + realpath
...
...     winpath3 = cext.GetFinalPathName(ntpath3, from_nt=True)
...     print(f'cext.GetFinalPathName({ntpath3!r}, from_nt=True) -> {winpath3!r}')
...
...     if realpath.startswith('\\\\'):
...         assert winpath2 == '\\\\?\\UNC\\' + realpath[2:]
...     else:
...         assert winpath3 == '\\\\?\\' + realpath
...
Path: 'C:\\'
os.path.realpath('C:\\') -> 'C:\\'
cext.RtlDosPathNameToNtPathName('C:\\') -> '\\??\\C:\\'
psutil._pswindows.convert_dos_path('\\??\\C:\\') -> 'C:\\'
cext.GetFinalPathName('C:\\', to_nt=True) -> '\\Device\\HarddiskVolume3\\'
cext.GetFinalPathName('\\??\\C:\\', from_nt=True, to_nt=True) -> '\\Device\\HarddiskVolume3\\'
cext.GetFinalPathName('\\??\\C:\\', from_nt=True) -> '\\\\?\\C:\\'
cext.GetFinalPathName('\\Device\\HarddiskVolume3\\', from_nt=True) -> '\\\\?\\C:\\'
Path: 'D:\\'
os.path.realpath('D:\\') -> 'D:\\'
cext.RtlDosPathNameToNtPathName('D:\\') -> '\\??\\D:\\'
psutil._pswindows.convert_dos_path('\\??\\D:\\') -> 'D:\\'
cext.GetFinalPathName('D:\\', to_nt=True) -> '\\Device\\HarddiskVolume7\\'
cext.GetFinalPathName('\\??\\D:\\', from_nt=True, to_nt=True) -> '\\Device\\HarddiskVolume7\\'
cext.GetFinalPathName('\\??\\D:\\', from_nt=True) -> '\\\\?\\D:\\'
cext.GetFinalPathName('\\Device\\HarddiskVolume7\\', from_nt=True) -> '\\\\?\\D:\\'
Path: 'G:\\'
os.path.realpath('G:\\') -> 'G:\\'
cext.RtlDosPathNameToNtPathName('G:\\') -> '\\??\\G:\\'
psutil._pswindows.convert_dos_path('\\??\\G:\\') -> 'G:\\'
cext.GetFinalPathName('G:\\', to_nt=True) -> '\\Device\\Volume{a4970278-c792-3e24-a0ee-12244993264d}\\'
cext.GetFinalPathName('\\??\\G:\\', from_nt=True, to_nt=True) -> '\\Device\\Volume{a4970278-c792-3e24-a0ee-12244993264d}\\'
cext.GetFinalPathName('\\??\\G:\\', from_nt=True) -> '\\\\?\\G:\\'
cext.GetFinalPathName('\\Device\\Volume{a4970278-c792-3e24-a0ee-12244993264d}\\', from_nt=True) -> '\\\\?\\G:\\'
Path: 'X:\\'
os.path.realpath('X:\\') -> 'D:\\venv'
cext.RtlDosPathNameToNtPathName('X:\\') -> '\\??\\X:\\'
psutil._pswindows.convert_dos_path('\\??\\X:\\') -> 'X:\\'
cext.GetFinalPathName('X:\\', to_nt=True) -> '\\Device\\HarddiskVolume7\\venv'
cext.GetFinalPathName('\\??\\X:\\', from_nt=True, to_nt=True) -> '\\Device\\HarddiskVolume7\\venv'
cext.GetFinalPathName('\\??\\X:\\', from_nt=True) -> '\\\\?\\D:\\venv'
cext.GetFinalPathName('\\Device\\HarddiskVolume7\\venv', from_nt=True) -> '\\\\?\\D:\\venv'
Path: 'Y:\\'
os.path.realpath('Y:\\') -> '\\\\localhost\\D$'
cext.RtlDosPathNameToNtPathName('Y:\\') -> '\\??\\Y:\\'
psutil._pswindows.convert_dos_path('\\??\\Y:\\') -> 'Y:\\'
cext.GetFinalPathName('Y:\\', to_nt=True) -> '\\Device\\Mup\\localhost\\D$'
cext.RtlDosPathNameToNtPathName('\\\\localhost\\D$') -> '\\??\\UNC\\localhost\\D$'
psutil._pswindows.convert_dos_path('\\??\\UNC\\localhost\\D$') -> '\\\\localhost\\D$'
cext.GetFinalPathName('\\??\\UNC\\localhost\\D$', from_nt=True, to_nt=True) -> '\\Device\\Mup\\localhost\\D$'
cext.GetFinalPathName('\\??\\Y:\\', from_nt=True, to_nt=True) -> '\\Device\\Mup\\localhost\\D$'
cext.GetFinalPathName('\\??\\Y:\\', from_nt=True) -> '\\\\?\\UNC\\localhost\\D$'
cext.GetFinalPathName('\\Device\\Mup\\localhost\\D$', from_nt=True) -> '\\\\?\\UNC\\localhost\\D$'
```

giampaolo commented:
Mmmm. I'm seeing little value in exposing `RtlDosPathNameToNtPathName` and `GetFinalPathName` just to use them in unit tests.

What I meant is: from unit tests call `convert_dos_path()` with different values so that you can test your own (new) implementation. 

Or would it make sense to expose the new C functions and have `convert_dos_path()` use them directly instead of this:

```python
    if rawdrive in {"\\??\\UNC", "\\Device\\Mup"}:
        rawdrive = '\\'.join(s.split('\\')[:4])
        driveletter = "\\\\" + s.split('\\')[3]
    elif rawdrive.startswith('\\??\\'):
        driveletter = s.split('\\')[2]
    else:
        driveletter = cext.QueryDosDevice(rawdrive)
```

...?

Are `RtlDosPathNameToNtPathName` and `GetFinalPathName` used in conjunctions the same as what you're doing here in python?

seekerFactory commented:
Handle


On Sun, Apr 20, 2025, 02:39 Giampaolo Rodola ***@***.***>
wrote:

> Mmmm. I'm seeing little value in exposing RtlDosPathNameToNtPathName and
> GetFinalPathName just to use them in unit tests.
>
> What I meant is: from unit tests call convert_dos_path() with different
> values so that you can test your own (new) implementation.
>
> Or would it make sense to expose the new C functions and have
> convert_dos_path() use them directly instead of this:
>
>     if rawdrive in {"\\??\\UNC", "\\Device\\Mup"}:
>         rawdrive = '\\'.join(s.split('\\')[:4])
>         driveletter = "\\\\" + s.split('\\')[3]
>     elif rawdrive.startswith('\\??\\'):
>         driveletter = s.split('\\')[2]
>     else:
>         driveletter = cext.QueryDosDevice(rawdrive)
>
> ...?
>
> Are RtlDosPathNameToNtPathName and GetFinalPathName used in conjunctions
> the same as what you're doing here in python?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/giampaolo/psutil/pull/2495#issuecomment-2816867448>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKALQ2SHW6F6V7CEWLHAKL22K3R5AVCNFSM6AAAAABURHXOICVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMJWHA3DONBUHA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
> *giampaolo* left a comment (giampaolo/psutil#2495)
> <https://github.com/giampaolo/psutil/pull/2495#issuecomment-2816867448>
>
> Mmmm. I'm seeing little value in exposing RtlDosPathNameToNtPathName and
> GetFinalPathName just to use them in unit tests.
>
> What I meant is: from unit tests call convert_dos_path() with different
> values so that you can test your own (new) implementation.
>
> Or would it make sense to expose the new C functions and have
> convert_dos_path() use them directly instead of this:
>
>     if rawdrive in {"\\??\\UNC", "\\Device\\Mup"}:
>         rawdrive = '\\'.join(s.split('\\')[:4])
>         driveletter = "\\\\" + s.split('\\')[3]
>     elif rawdrive.startswith('\\??\\'):
>         driveletter = s.split('\\')[2]
>     else:
>         driveletter = cext.QueryDosDevice(rawdrive)
>
> ...?
>
> Are RtlDosPathNameToNtPathName and GetFinalPathName used in conjunctions
> the same as what you're doing here in python?
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/giampaolo/psutil/pull/2495#issuecomment-2816867448>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AAKALQ2SHW6F6V7CEWLHAKL22K3R5AVCNFSM6AAAAABURHXOICVHI2DSMVQWIX3LMV43OSLTON2WKQ3PNVWWK3TUHMZDQMJWHA3DONBUHA>
> .
> You are receiving this because you are subscribed to this thread.Message
> ID: ***@***.***>
>

klightspeed commented:
Perhaps I could remove `RtlDosPathNameToNtPathName`, since it's basically only doing `'\\??\\' + path` for DOS paths and `'\\??\\UNC\\' + path[2:]` for UNC paths.

`GetFinalPathName` does need to be able to access the path to convert it, and its DOS path would need the `\\?\` or `\\?\UNC\` prefix removed.

Perhaps instead of using `GetFinalPathName` to convert a DOS path to a NT path, I could mock `QueryDosDevice` to return a given drive letter given a NT device path.

giampaolo commented:
@klightspeed: I didn't realize this had a failure. Could you please fix it?

```
2025-04-29T18:38:48.6315448Z __________________ TestSystemAPIs.test_convert_dos_path_unc ___________________
2025-04-29T18:38:48.6316118Z psutil\tests\test_windows.py:288: in test_convert_dos_path_unc
2025-04-29T18:38:48.6316615Z     assert psutil._pswindows.convert_dos_path(ntpath1) == winpath
2025-04-29T18:38:48.6317175Z E   AssertionError: assert '\\C$\\Windows\\Temp' == '\\\\localhos...Windows\\Temp'
2025-04-29T18:38:48.6317614Z E     - \\localhost\C$\Windows\Temp
2025-04-29T18:38:48.6317906Z E     + \C$\Windows\Temp
2025-04-29T18:38:48.6320264Z 
```

klightspeed commented:
> @klightspeed: I didn't realize this had a failure. Could you please fix it?
> 
> ```
> 2025-04-29T18:38:48.6315448Z __________________ TestSystemAPIs.test_convert_dos_path_unc ___________________
> 2025-04-29T18:38:48.6316118Z psutil\tests\test_windows.py:288: in test_convert_dos_path_unc
> 2025-04-29T18:38:48.6316615Z     assert psutil._pswindows.convert_dos_path(ntpath1) == winpath
> 2025-04-29T18:38:48.6317175Z E   AssertionError: assert '\\C$\\Windows\\Temp' == '\\\\localhos...Windows\\Temp'
> 2025-04-29T18:38:48.6317614Z E     - \\localhost\C$\Windows\Temp
> 2025-04-29T18:38:48.6317906Z E     + \C$\Windows\Temp
> 2025-04-29T18:38:48.6320264Z 
> ```

It looks like for older pythons I need to add a backslash to the `rawdrive`

```
Python 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> s = '\\??\\UNC\\localhost\\C$\\Windows\\Temp'
>>> rawdrive = '\\'.join(s.split('\\')[:4])
>>> driveletter = "\\\\" + s.split('\\')[3]
>>> rawdrive
'\\??\\UNC\\localhost'
>>> driveletter
'\\\\localhost'
>>> remainder = s[len(rawdrive) :]
>>> remainder
'\\C$\\Windows\\Temp'
>>> os.path.join(driveletter, remainder)
'\\C$\\Windows\\Temp'
```

```
Python 3.13.1 (tags/v3.13.1:0671451, Dec  3 2024, 19:06:28) [MSC v.1942 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> s = '\\??\\UNC\\localhost\\C$\\Windows\\Temp'
>>> rawdrive = '\\'.join(s.split('\\')[:4])
>>> driveletter = "\\\\" + s.split('\\')[3]
>>> rawdrive
'\\??\\UNC\\localhost'
>>> driveletter
'\\\\localhost'
>>> remainder = s[len(rawdrive) :]
>>> remainder
'\\C$\\Windows\\Temp'
>>> os.path.join(driveletter, remainder)
'\\\\localhost\\C$\\Windows\\Temp'
```

```
Python 3.8.10 (tags/v3.8.10:3d8993a, May  3 2021, 11:48:03) [MSC v.1928 64 bit (AMD64)] on win32
Type "help", "copyright", "credits" or "license" for more information.
>>> import os
>>> s = '\\??\\UNC\\localhost\\C$\\Windows\\Temp'
>>> rawdrive = '\\'.join(s.split('\\')[:4]) + '\\'
>>> driveletter = "\\\\" + s.split('\\')[3]
>>> remainder = s[len(rawdrive) :]
>>> os.path.join(driveletter, remainder)
'\\\\localhost\\C$\\Windows\\Temp'
```

```
(psutil) D:\venv\psutil\psutil>python --version
Python 3.13.1

(psutil) D:\venv\psutil\psutil>git describe
release-6.1.1-95-gfe4742a6

(psutil) D:\venv\psutil\psutil>python -m pytest -v "psutil\tests\test_windows.py::TestSystemAPIs::test_convert_dos_path_drive" "psutil\tests\test_windows.py::TestSystemAPIs::test_convert_dos_path_unc"
================================================= test session starts =================================================
collected 2 items

psutil/tests/test_windows.py::TestSystemAPIs::test_convert_dos_path_drive PASSED
psutil/tests/test_windows.py::TestSystemAPIs::test_convert_dos_path_unc PASSED

================================================== 2 passed in 0.30s ==================================================

```
